excellent/types/base.go

```diff
@@ -0,0 +1,262 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/nyaruka/goflow/utils"
+
+	"github.com/shopspring/decimal"
+)
+
+func init() {
+	decimal.MarshalJSONWithoutQuotes = true
+}
+
+// XType represents a type in Excellent
+type XType int
+
+// the supported types
+const (
+	XTypeString XType = iota
+	XTypeNumber
+	XTypeBool
+	XTypeTime
+	XTypeArray
+	XTypeObject
+	XTypeError
+	XTypeNil
+)
+
+// XValue is the base interface of all Excellent types
+type XValue interface {
+	Type() XType
+	ToJSON() XString
+}
+
+// XPrimitive is the base interface of all Excellent primitive types
+type XPrimitive interface {
+	XValue
+
+	ToString() XString
+	ToBool() XBool
```

Comment:
`ToString` and `ToBool` are somewhat different to other conversions since every type should be able to render itself as these without blowing up